float64 ed 
float64 eang
